### PR TITLE
Segmentation head

### PR DESCRIPTION
**Pull request by alexrey88**
Adding segmentation head as an auxiliary task is the purpose of this PR. :)

### CODE REVIEW

**Comment by alexrey88** [comment]
All right. Segmentation head is running well. The results are not as good as I thought they would be but maybe there's something I've missed in the code, or maybe I should play with the learning rate/batch_size more. Here's the current running experiment : [here](https://www.comet.ml/alexrey88/omnigan/c9c24ccb7233428197c2960f2529156d?experiment-tab=images&groupBy=false&orderBy=desc&search=val_s_s&sortBy=name). Waiting for you reviews !! :)

File: omnigan/utils.py [hunk 1]
```diff
@@ -287,7 +283,7 @@ def make_json_file(
     keys,
     addresses,  # for windows user, use "\\" instead of using "/"
     name_of_the_json_file="jsonfile.json",
-    splitter = '/'
```
Comment (melisandeteng) [inline omnigan/utils.py:290]: Ooooh this was not "black-ified "!

File: omnigan/utils.py [hunk 2]
```diff
@@ -338,6 +334,59 @@ def make_json_file(
...
+            file_found = False
+            for file_path in files:
+                if file_name in file_path:
+                    file_found = True
+                    new_ims_list[i][new_task_name] = file_path
+                    break
+            if file_found:
+                break
+            else:
+                print("Error! File ", file_name, "not found in directory!")
```
Comment (melisandeteng) [inline omnigan/utils.py:383]: oh  so does that mean we don't write the file if we have one seg map missing for example ? and then we have to handle that "by hand " right ?

File: omnigan/data.py [hunk 3]
```diff
@@ -22,8 +22,60 @@
...
+        1: [55, 55, 55, 255],  # Ground
+        2: [0, 255, 255, 255],  # Building
+        3: [255, 212, 0, 255],  # Traffic items
+        4: [0, 255, 0, 255],  # Vegetation
+        5: [255, 97, 0, 255],  # Terrain
+        6: [255, 0, 0, 255],  # Car
+        7: [0, 255, 0, 255],  # Trees
+        8: [220, 20, 60, 255],  # Person
+        9: [8, 19, 49, 255],  # Sky
+        10: [0, 80, 100, 255],  # Default
```
[the remark this answers is not in the document]
Comment (melisandeteng) [reply omnigan/data.py:50]: It's just because you didn't indicate any 255, 255, 255, 255 class

[the remark this answers is not in the document]
Comment (melisandeteng) [reply omnigan/data.py:50]: But maybe we don't need to ? But then what do you map the unknown objects to ? What I'm saying is, maybe there should be a "garbage" class ?

File: omnigan/data.py [hunk 4]
```diff
@@ -66,6 +118,90 @@ def decode_segmap(image, nc=19):
...
+    """Takes a pixel as input and finds the closest known pixel value corresponding to a class in dict_classes
+    Arguments:
+        pixel -- tuple pixel (R,G,B,A)
+    Returns:
+        tuple pixel (R,G,B,A) corresponding to a key (a class) in dict_classes
+    """
+    min_dist = float("inf")
+    closest_pixel = None
+    for pixel_value in dict_classes.keys():
+        dist = np.sum(np.absolute(np.subtract(pixel, pixel_value)))
```
[the remark this answers is not in the document]
Comment (melisandeteng) [reply omnigan/data.py:129]: yes, right, I don't know what I was thinking... but maybe Euclidean distance is better ? 
That being said, I doubt there will actually be that significant changes so that you really have a random color that doesn't correspond to anything, right ?
 I guess an extra thing would then be to look at neighbouring classes to decide in the case of a tie but that seems like an overkill, so I don't think you need to do that!

File: omnigan/utils.py (hunk 2 above)
Comment (melisandeteng) [reply omnigan/utils.py:383]: Yes I don't know what the best practice is, actually.  Sometimes what I do is continue the process and write in a separate file the files that couldn't be handled. But given the arguments the function takes, I'm not sure if it would be really better in your case

File: omnigan/data.py [hunk 5]
```diff
@@ -22,8 +22,54 @@
...
+    Returns:
+        RGB tensor of size (1) x (3) x (H) x (W)
+    # """
+
+    if is_target:  # Target is size 1 x 1 x H x W
+        idx = tensor.squeeze(0).squeeze(0)
+    else:  # Prediction is size 1 x nc x H x W
+        idx = torch.argmax(tensor.squeeze(0), dim=0)
+
+    indexer = torch.tensor(list(classes_dict[domain].values()))[:, :3]
```
Comment (vict0rsch) [inline omnigan/data.py:68]: inspired by that question https://stackoverflow.com/questions/46868855/numpy-replacing-values-in-a-2d-array-efficiently-using-a-dictionary-as-a-map the implementation I pushed should be ~50-100 times faster

Comment (alexrey88) [reply omnigan/data.py:68]: Damn, cool! It was indeed slow haha. :)

File: omnigan/data.py [hunk 6]
```diff
@@ -22,8 +22,54 @@
...
+    return indexer[idx].permute(2, 0, 1).to(torch.float32).unsqueeze(0)
```
[the remark this answers is not in the document]
Comment (alexrey88) [reply omnigan/data.py:69]: Hmmmm well I think it's ok ? This function is to have a RGB image for visualisation on comet, and our "x" task and the others are float tensor too, so I think it's fine.

File: omnigan/data.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (melisandeteng) [reply omnigan/data.py:124]: Yes I guess I would just put Euclidian distance for the sake of geometers :)

File: omnigan/data.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (melisandeteng) [reply omnigan/data.py:49]: Right, I think  what I wanted to say (and didn't say) was that I thought there was a 255,255,255,0 case in the simulated datan abd wanted to know how you dealt, with that, if you had mapped it to 255,255,255,255 ?

[the remark this answers is not in the document]
Comment (alexrey88) [reply omnigan/data.py:49]: Hmmm wow I don't remember this issue with [255,255,255,0] pixels.. But yeah it would be mapped to 255,255,255,255.
